ghcide/src/Development/IDE/Core/Tracing.hs

```diff
@@ -71,36 +72,46 @@ otTracedAction key file success act = actionBracket
         unless (success res) $ setTag sp "error" "1"
         return res)
 
-startTelemetry :: Logger -> Var Values -> IO ()
-startTelemetry logger stateRef = do
+startTelemetry :: Bool -> Logger -> Var Values -> IO ()
+startTelemetry allTheTime logger stateRef = do
     instrumentFor <- getInstrumentCached
     mapCountInstrument <- mkValueObserver "values map count"
 
-    _ <- regularly (1 * seconds) $
-        withSpan_ "Measure length" $
-        readVar stateRef
-        >>= observe mapCountInstrument . length
-
-    _ <- regularly (1 * seconds) $ do
-        values <- readVar stateRef
-        let keys = nub
-                     $ Key GhcSession : Key GhcSessionDeps
-                     : [ k | (_,k) <- HMap.keys values
-                           -- do GhcSessionIO last since it closes over stateRef itself
-                           , k /= Key GhcSessionIO]
-                     ++ [Key GhcSessionIO]
-        !groupedForSharing <- evaluate (keys `using` seqList r0)
-        measureMemory logger [groupedForSharing] instrumentFor stateRef
-          `catch` \(e::SomeException) ->
-            logInfo logger ("MEMORY PROFILING ERROR: " <> fromString (show e))
-    return ()
+    installSigUsr1Handler $ do
+        logInfo logger "SIGUSR1 received: performing memory measurement"
+        performMeasurement logger stateRef instrumentFor mapCountInstrument
+
+    when allTheTime $ void $ regularly (1 * seconds) $
+        performMeasurement logger stateRef instrumentFor mapCountInstrument
   where
         seconds = 1000000
 
         regularly :: Int -> IO () -> IO (Async ())
         regularly delay act = async $ forever (act >> threadDelay delay)
 
-{-# ANN startTelemetry ("HLint: ignore Use nubOrd" :: String) #-}
+
+performMeasurement ::
+  Logger ->
+  Var (HMap.HashMap (NormalizedFilePath, Key) (Value Dynamic)) ->
+  (Maybe Key -> IO OurValueObserver) ->
+  Instrument 'Asynchronous a m' ->
+  IO ()
+performMeasurement logger stateRef instrumentFor mapCountInstrument = do
+    withSpan_ "Measure length" $ readVar stateRef >>= observe mapCountInstrument . length
+
+    values <- readVar stateRef
+    let keys = nub
```

Comment:
No way an O(n^2) algorithm is ever a good idea. If this can't be nubOrd, make it nubHash and write that (if we don't have it somewhere already)

Reply:
Good catch! I thought I had replaced this nub with nubOrd. And this is where the `Ord` requirement comes from.